### ISSUE TITLE
Instability in Megatron Training

### ISSUE THREAD

**Issue by blackPython** [modelscope/ms-swift#7906]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new question or discussion topic. / 我已经搜索过现有的 issues，确认这是一个新的问题与讨论。

### Question Description / 问题描述

I have tried supervised finetuning Qwen3 32B using both megatorn and deepspeed. I have found that megatron fails due to explosion in loss, where as deepspeed is stable. I have used the same hyper-parameters in both the experiments. Below are the yaml configs files:
```
#####   DeepSpeed Parameters ######
# Model
model: "Qwen/Qwen3-32B"
use_hf: true

# Training Strategy
train_type: "lora"
lora_rank: 256
lora_alpha: 256
lora_dropout: 0.05
target_modules: "all-linear"

# Optimization
learning_rate: 2e-4
weight_decay: 0.0
max_grad_norm: 0.5
warmup_ratio: 0.05
num_train_epochs: 1
per_device_train_batch_size: 1
gradient_accumulation_steps: 1 # Effective global batch size = 8 nodes * 8 GPUs/node = 64
adam_beta2: 0.999
lr_scheduler_type: "cosine"

# Mixed Precision
bf16: true

# Flash Attention
attn_impl: "flash_attn"

# DeepSpeed
deepspeed: "zero3"

# Gradient Checkpointing (replaces Megatron recompute_granularity=full)
gradient_checkpointing: true

#####   Meagtron Parameters ######

# model
model: "Qwen/Qwen3-32B"
use_hf: true
merge_lora: true
load_safetensors: true
save_safetensors: true

# Training Strategy
train_type: "lora"         # lora or full
lora_rank: 256
lora_alpha: 256
lora_dropout: 0.05
target_modules: "all-linear"
finetune: true
weight_decay: 0.0

# Optimization
lr: 1e-4
min_lr: 0
lr_warmup_fraction: 0.05
clip_grad: 0.5
max_epochs: 1
micro_batch_size: 4
global_batch_size: 64
adam_beta2: 0.999

# Parallelism
tensor_model_parallel_size: 4
context_parallel_size: 2
pipeline_model_parallel_size: 1
sequence_parallel: true

# MoE Settings (will be ignored if not using MoE)
moe_permute_fusion: true
moe_grouped_gemm: true
moe_shared_expert_overlap: true
moe_aux_loss_coeff: 0

# Memory / Recompute
recompute_granularity: "full"
recompute_method: "uniform"
recompute_num_layers: 1

# Misc
attention_backend: "flash"
cross_entropy_loss_fusion: true
``` 

<img width="1423" height="856" alt="Image" src="https://github.com/user-attachments/assets/0da08949-c314-492d-a169-da2397135d5e" />
<img width="1423" height="856" alt="Image" src="https://github.com/user-attachments/assets/0a7de645-1c53-495b-be1c-1077b23c55fd" />

You can from the loss curves that megatron training is unstable (brown loss curve). I am using the latest versions of ms-swift `3.12.x`. 

Does anyone why this is happening? I have tried setting `cross_entropy_loss_fusion: false` it does not help. what helps is reducing learning rate to `1e-4`, but then reducing the LR means I have to train longer.

Any help is really appreciated.

**Comment by blackPython** [comment]
Any update on this? I am happy to help you guys out, if you have any clue where to look for the potential bug or how to look for the bug

**Comment by blackPython** [comment]
HI @Jintao-Huang, Do we have any updates on this?

**Comment by Jintao-Huang** [comment]
Try the release/3.12 or main branch and see if it works.

**Comment by blackPython** [comment]
Using the 3.12.4 resolved this issue.
